src/main/java/reposense/report/ReportGenerator.java

```diff
@@ -173,38 +180,52 @@ private static void setLandingPageTitle(String filePath, String pageTitle) throw
 
     /**
      * Clone, analyze and generate the report for repositories in {@code repoLocationMap}.
-     * Performs analysis and report generation of each repository in parallel with the cloning of the next repository.
+     * Performs cloning, analysis and report generation of each repository in parallel.
      *
      * @return A list of paths to the JSON report files generated for each repository.
      */
     private static List<Path> cloneAndAnalyzeRepos(List<RepoConfiguration> configs, String outputPath) {
         Map<RepoLocation, List<RepoConfiguration>> repoLocationMap = groupConfigsByRepoLocation(configs);
-        RepoCloner repoCloner = new RepoCloner();
-        RepoLocation clonedRepoLocation = null;
-
         List<RepoLocation> repoLocationList = new ArrayList<>(repoLocationMap.keySet());
 
-        RepoLocation currRepoLocation = repoLocationList.get(0);
-        repoCloner.cloneBare(repoLocationMap.get(currRepoLocation).get(0));
-
-        List<Path> generatedFiles = new ArrayList<>();
-        for (int index = 1; index <= repoLocationList.size(); index++) {
-            RepoLocation nextRepoLocation = (index < repoLocationList.size()) ? repoLocationList.get(index) : null;
-            clonedRepoLocation = repoCloner.getClonedRepoLocation();
-
-            // Clones the next location while analyzing the previously cloned repos in parallel.
-            if (nextRepoLocation != null) {
-                repoCloner.cloneBare(repoLocationMap.get(nextRepoLocation).get(0));
-            }
-
-            if (clonedRepoLocation == null) {
-                handleCloningFailed(configs, currRepoLocation);
-            } else {
-                generatedFiles.addAll(analyzeRepos(outputPath, configs, repoLocationMap.get(clonedRepoLocation),
-                        repoCloner.getCurrentRepoDefaultBranch()));
+        ExecutorService executor = Executors.newFixedThreadPool(4);
```

Comment:
Would it be more ideal to adapt this number to match the number of available threads that the user's machine has? Something along the lines of:

```java
int threads = Runtime.getRuntime().availableProcessors();
ExecutorService executor = Executors.newFixedThreadPool(threads);
```

We should also impose an upper limit so that we don't open too many Git clones at once coming from a single host, so that we do not accidentally trigger any auto throttling by GitHub.

Reply:
Agree with @dcshzj suggestion.

Reply:
> 
> 
> Would it be more ideal to adapt this number to match the number of available threads that the user's machine has? Something along the lines of:
> 
> ```java
> int threads = Runtime.getRuntime().availableProcessors();
> ExecutorService executor = Executors.newFixedThreadPool(threads);
> ```
Agreed, this makes more sense.

I tried this out; it seems that if the CPU supports hyper-theading, the number of "available processors" is doubled, but the performance increase is more marginal. On my PC, with 8 threads, it took 1 min 8 sec to generate the report for the CS2113 AY2019/20 Sem 2 group project (17 repos). This is a slight improvement over the 1 min 20 sec taken with 4 threads. For reference, vanilla RepoSense takes about 2 min 40 sec.

>
> 
> We should also impose an upper limit so that we don't open too many Git clones at once coming from a single host, so that we do not accidentally trigger any auto throttling by GitHub.

According to [their docs](https://docs.github.com/en/github/getting-started-with-github/troubleshooting-connectivity-problems#troubleshooting-slow-downloads-and-intermittent-slow-connections), GitHub does not throttle bandwidth per user. But I think we could still end up with bandwidth issues with many simultaneous connections. How many connections do you think is a suitable upper limit?

Reply:
Sorry I missed this, I don't have an exact figure to suggest but I think 4 threads should be a good maximum.